### PR TITLE
Refactor doctor and make sure issues with sbt are properly reported

### PR DESCRIPTION
**Pull request by tgodzik**
The main goal of this refactor is to make sure we have just one category of problems that will be reported both in the message request and the doctor. Previously, both would need to be updated independently and testing for those was scarce. Thanks to having a single category of problems we can now properly unit test them to make sure every message is sensible. It's also easier now to see what each problem is caused by and what message will be displayed to the user.

Also, this should fix https://github.com/scalameta/metals/issues/2167 - sbt issues will now be reported separately, so that the users are not confused.

### CODE REVIEW

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala [hunk 1]
```diff
@@ -39,6 +38,13 @@ final class Doctor(
   private val hasProblems = new AtomicBoolean(false)
   private var bspServerName: Option[String] = None
   private var bspServerVersion: Option[String] = None
+  private val problemResolver =
+    new ProblemResolver(
+      workspace,
+      () => bspServerVersion.exists(isUnsupportedBloopVersion),
+      () => bspServerVersion,
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Doctor.scala:45]: I think the logic is a bit mixed in here and in `ProblemResolver.scala` and it makes it a _bit_ hard to follow.

Do you think it's necessary to actually have `isUnsupportedBloopVersion` _and_ `bloopVersion` be parameters in `ProblemResolver`? What do think of instead just move all of this logic into `ProblemResolver` and handle it all there. We'd move the `isUnsupportedBloopVersion` method right below this into there as well. Then there would be no need to have both of these parameter, just the `BloopVersion`, and then we could do the logic of determining if it's unsupported in the `ProblemResolver`. wdyt?

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+import scala.meta.internal.metals.Messages
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.metals.ScalaTarget
+import scala.meta.internal.metals.ScalaVersions
+import scala.meta.io.AbsolutePath
+
+class ProblemResolver(
+    workspace: AbsolutePath,
+    isUnsupportedBloopVersion: () => Boolean,
+    bloopVersion: () => Option[String],
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:14]: Mentioned this in `Doctor.scala`, but I think it might be cleaner to only have `bloopVersion` here and then move the `isUnsupportedBloopVersion` method from `Doctor.scala` into here and have the logic to determine whether or not it's unsupported in here since a lot of the other logic for finding various problems is in here.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 3]
```diff
@@ -0,0 +1,75 @@
+package scala.meta.internal.troubleshoot
+
+import scala.meta.internal.metals.BuildInfo
+import scala.meta.internal.metals.ScalaVersions
+
+sealed abstract class ScalaProblem {
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:6]: More of just a curiosity, but I notice in a lot of Scalameta projects we prefer `abstract class` over `trait`. What's the main reason we do that?

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  def message: String
+
+  protected val hint = "run 'Build import' to enable code navigation."
+
+}
+
+// scala version issues
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:13]: I think we can probably just remove this comment or include proper scaladocs for these.

Also related, if possible can you add a short Scaladoc to `ScalaProblem`. I find that when all of the classes have short descriptions it makes it much easier for someone new to the codebase to get a quick idea of what a class is meant for then.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+      val recommended = ScalaVersions.recommendedVersion(version)
+      s"Upgrade to Scala $recommended and " + hint
+    } else {
+      val versionToUpgradeTo =
+        if (ScalaVersions.isScala3Version(version)) {
+          s"Scala ${BuildInfo.scala3}"
+        } else {
+          s"Scala ${BuildInfo.scala213} or ${BuildInfo.scala212}"
+        }
+      s"Code navigation is not supported for this compiler version, change to " + versionToUpgradeTo + " and " + hint
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:26]:
```suggestion
      s"Code navigation is not supported for this compiler version, please change to " + versionToUpgradeTo + " and " + hint
```

Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:26]: Make sure Metals is polite 😄

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+case class DeprecatedScalaVersion(version: String) extends ScalaProblem {
+  override def message: String =
+    s"Scala $version might not be supported in upcoming versions of Metals, " +
+      s"please upgrade to Scala ${ScalaVersions.recommendedVersion(version)}."
+}
+
+case class FutureScalaVersion(version: String) extends ScalaProblem {
+  override def message: String = s"Scala $version is not yet supported"
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:38]: Should we also include more info to downgrade to the latest supported version? Similar to what we have in `DeprecatedScalaVersion`?
```suggestion
  override def message: String = s"Scala $version is not yet supported, " +
    s"please downgrade to Scala ${ScalaVersions.recommendedVersion(version)}."
```

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// semanticdb problems
+case class SemanticDBDisabled(
+    scalaVersion: String,
+    bloopVersion: String,
+    isBloopVersionSupported: Boolean
+) extends ScalaProblem {
+  override def message: String = {
+    if (isBloopVersionSupported) {
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:48]: This confuses me a bit? I'd expect that `isBloopVersionSupported` is true, then we'd have no issue. But this message is saying that it's not supported and Metals requires at least Bloop x. Is this correct?

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+      s"""|The installed Bloop server version is $bloopVersion while Metals requires at least Bloop version ${BuildInfo.bloopVersion},
+          |To fix this problem please update your Bloop server.""".stripMargin
+    } else if (ScalaVersions.isSupportedScalaVersion(scalaVersion)) {
+      hint.capitalize
+    } else {
+      "Enable semanticdb plugin for the project."
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:54]: I'm not 100% sure on the situation on when this will be hit, but I think the more info we can give here the better, since for a lot of people this may not be clear. They might want to why this is being shown. Maybe this will just be combined with other messages, and if so ignore this, but I think having a bit more info, something like "Semanticdb is required for code navigation to work correctly in your project, however the semanticdb plugin doesn't seem to be enabled for your project. Please enable the semanticdb plugin for this project in order for code navigation to work correctly." wdyt?

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 9]
```diff
@@ -0,0 +1,75 @@
...
+}
+case class MissingSourceRoot(sourcerootOption: String) extends ScalaProblem {
+  override def message: String =
+    s"Add the compiler option $sourcerootOption to ensure code navigation works."
+}
+
+// sbt problems
+case object UnsupportedSbtVersion extends ScalaProblem {
+  override def message: String =
+    "Code navigation is not supported for this sbt version, please upgrade to 1.3.2 and reimport the build"
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:66]:
```suggestion
    "Code navigation is not supported for this sbt version, please upgrade to at least 1.3.2 and reimport the build"
```

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala [hunk 10]
```diff
@@ -0,0 +1,75 @@
...
+}
+case object DeprecatedSbtVersion extends ScalaProblem {
+  override def message: String =
+    "Code navigation might not be supported in the future for this sbt version, please upgrade to 1.3.2 and reimport the build"
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:70]:
```suggestion
    "Code navigation might not be supported in the future for this sbt version, please upgrade to at least 1.3.2 and reimport the build"
```

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala [hunk 11]
```diff
@@ -0,0 +1,161 @@
...
+      issue <- findProblem(target)
+    } yield {
+      issue match {
+        case UnsupportedScalaVersion(version) => unsupportedVersions += version
+        case DeprecatedScalaVersion(version) => deprecatedVersions += version
+        case FutureScalaVersion(version) => futureVersions += version
+        case _: SemanticDBDisabled => misconfiguredProjects += 1
+        case _: MissingSourceRoot => misconfiguredProjects += 1
+        case UnsupportedSbtVersion => unsupportedSbt = true
+        case DeprecatedSbtVersion => deprecatedSbt = true
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:45]: For some of these, it seems like we have an extra middle step that isn't really necessary. For example for these, we find the issue, and if it's one of these two, we set it to true, then down below we check if it's true and create a Message. Couldn't we just skip the middle step and instead here instead of returning a Boolean return `Some(message)` or None right away and then there is no need to check for it down below?

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+      case single :: Nil => Some(single)
+      case Nil => None
+      case messages if messages == scalaVersionsMessages =>
+        Some(
+          s"Your build definition contains multiple unsupported and deprecated Scala versions."
+        )
+      case _ =>
+        Some(
+          s"Multiple problems detected in your build."
+        )
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:121]: So in this scenario, we do all the work up above to potentially get multiple messages, but then we never show them to the user, we just show `Multiple problems detected in your build`? Wouldn't it be better to actually show them to the user? I'd assume that `problemSummary` would include the summary of issues, not just a message saying there is issues.

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala [hunk 13]
```diff
@@ -126,7 +132,7 @@ final class Doctor(
   def check(serverName: String, serverVersion: String): Unit = {
     bspServerName = Some(serverName)
     bspServerVersion = Some(serverVersion)
-    val summary = problemSummary
+    val summary = problemResolver.problemSummary(allTargets())
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Doctor.scala:135]: Mentioned this at the definition site, but I think this is a bit misleading since `summary` here isn't really a summary, but just a message about issues. What do you think of actually including all of those messages if there are more than one here?

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala [hunk 14]
```diff
@@ -587,6 +587,29 @@ object Messages {
     }
   }
 
+  object DeprecatedSbtVersion {
+    def message: String = {
+      val recommended = "1.3.2"
+      s"You are using an old sbt version, navigation for which might not be supported in the future versions of Metals. " +
+        s"Please upgrade at least to sbt $recommended."
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Messages.scala:594]:
```suggestion
        s"Please upgrade to at least sbt $recommended."
```
Just a nit. I'm not sure why, but I think this sounds more natural.

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala [hunk 15]
```diff
@@ -587,6 +587,29 @@ object Messages {
...
+    }
+  }
+
+  object UnsupportedSbtVersion {
+    def message: String = {
+      val recommended = "1.3.2"
+      s"You are using an old sbt version, navigation for which is not supported in this version of Metals. " +
+        s"Please upgrade at least to sbt $recommended."
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Messages.scala:602]:
```suggestion
        s"Please upgrade to at least sbt $recommended."
```

**Review by ckipp01 — CHANGES_REQUESTED** [review]
As always, great job on this. I have a few questions/comments. Let me know what you think.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 3 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala:6]: Haha, no idea, I didn't even realize I created an abstract class. :sweat_smile:

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala (hunk 12 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:121]: So problem summary will be shown as a message to the user and if we get too much problems then it will be quite bloated. 

`Multiple problems detected in your build` will be accompanied by `More information`, so people can easily take a look. Though I have never honestly seen more than one issue at the same time, so it's highly improbable scenario.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala (hunk 11 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:45]: I think it's easier to understand the way it is. First we check what problems we have and only then translate to the messages. Otherwise we will have to mix those two steps, but I do understand what you mean and I can change it if you feel like it will be better.

Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:45]: I get what you mean. For consistency then feel free to just leave it as is.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala (hunk 12 above)
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ProblemResolver.scala:121]: Ah ok I totally missed the `More information` part. If that's there, then nevermind.

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala (hunk 1 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/Doctor.scala]: Good idea! I moved everything to `ProblemResolver` I also changed to only use `() => Option[BspSession]` which replaces 3 parameters in the doctor and we no longer need the vars.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 4 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: I removed the smaller comments and add scaladoc to `ScalaProblem`

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 5 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: Fixed! :sweat_smile:

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 6 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: Good idea, fixed!

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 7 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: Och yeah, that was actually badly named. Renamed to `unsupportedBloopVersion`

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 8 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: That should almost never happen, but your message is a much better one for sure.

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 9 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: Fixed!

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 10 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: Fixed!

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala (hunk 13 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/Doctor.scala]: I renamed it to `problemMessage`, but I think we should stay with a single message, since it avoids spamming the user and the `More information` option should be visible enough for people to easily check what is going on.

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala (hunk 14 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/Messages.scala]: Yeah, sounds better! Changed!

File: metals/src/main/scala/scala/meta/internal/metals/Messages.scala (hunk 15 above)
Comment (tgodzik) [reply metals/src/main/scala/scala/meta/internal/metals/Messages.scala]: Changed!

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala (hunk 1 above)
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/metals/Doctor.scala]: win, win, win 👍🏼

File: metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala (hunk 7 above)
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/troubleshoot/ScalaProblem.scala]: haha I'm glad I caught this then.

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala (hunk 13 above)
Comment (ckipp01) [reply metals/src/main/scala/scala/meta/internal/metals/Doctor.scala]: yea sounds good 👍🏼

File: metals/src/main/scala/scala/meta/internal/metals/Doctor.scala [hunk 16]
```diff
@@ -30,22 +29,19 @@ final class Doctor(
...
-    currentBuildServer: () => Option[String],
+    currentBuildServer: () => Option[BspSession],
     calculateNewBuildServer: () => BspResolvedResult,
     httpServer: () => Option[MetalsHttpServer],
     tables: Tables,
     clientConfig: ClientConfiguration
 )(implicit ec: ExecutionContext) {
   private val hasProblems = new AtomicBoolean(false)
-  private var bspServerName: Option[String] = None
-  private var bspServerVersion: Option[String] = None
```
Comment (ckipp01) [inline metals/src/main/scala/scala/meta/internal/metals/Doctor.scala:41]: Nice to see these are needed anymore 🦾

**Review by ckipp01 — APPROVED** [review]
This looks good @tgodzik! Hopefully this will help people diagnose what's going on! LGTM!
